### PR TITLE
Fix dead lock might happen in shutdown hook.

### PR DESCRIPTION
**Pull request by beeender**
A simple framework is created for multi-processes testing.
And since it can be used to verify #1328 , merge it earlier.

### CODE REVIEW

**Comment by beeender** [comment]
@realm/java Help to review. Thanks!

File: realm/src/androidTest/java/io/realm/RealmInterprocessTest.java [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+    }
+
+    @Override
+    protected void setUp() throws Exception {
+        super.setUp();
+
+        Realm.deleteRealm(new RealmConfiguration.Builder(getContext()).build());
+
+        // Start the testing service
+        serviceLatch = new CountDownLatch(1);
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmInterprocessTest.java]: perhaps rename this to serviceStartedLatch?

File: realm/src/androidTest/java/io/realm/RealmInterprocessTest.java [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+        Intent intent = new Intent(getContext(), InterprocessService.class);
+        getContext().bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE);
+        // Wait until service started
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmInterprocessTest.java]: Then this comment could be deleted :)

File: realm/src/androidTest/java/io/realm/RealmInterprocessTest.java [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    protected void tearDown() throws Exception {
+        if (testRealm != null) {
+            testRealm.close();
+        }
+
+        getContext().unbindService(serviceConnection);
+        serviceMessenger = null;
+
+        // Wait until the service process terminated.
+        while (getServiceInfo() != null) {
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmInterprocessTest.java]: This looks a bit dangerous. I would add a counter and throw an exception after 10 tries or something

File: realm/src/androidTest/java/io/realm/RealmInterprocessTest.java [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+            if (InterprocessService.class.getName().equals(service.service.getClassName())) {
+                return service;
+            }
+        }
+        return null;
+    }
+
+    // Our interprocess testing rely on the System.exit(0) call! So if this one fails, all other testing will fail.
+    // Always keep this one as the top one test case!
+    // A. Open a realm, close it, then call Runtime.getRuntime().exit(0).
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmInterprocessTest.java]: Either A/B or 1/2

File: realm/src/androidTest/java/io/realm/RealmInterprocessTest.java [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        Looper.loop();
+    }
+
+    // 1. Main process create Realm, write one object.
+    // A. Service process open Realm, check if there is one and only one object.
+    public void testCreateInitialRealm() throws InterruptedException {
+        new InterprocessHandler(new Runnable() {
+            @Override
+            public void run() {
+                // Step 1
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmInterprocessTest.java]: 1/2 or A/B

File: realm/src/androidTest/java/io/realm/RealmInterprocessTest.java [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+                testRealm = Realm.getInstance(getContext());
+                assertEquals(testRealm.allObjects(AllTypes.class).size(), 0);
+                testRealm.beginTransaction();
+                testRealm.createObject(AllTypes.class);
+                testRealm.commitTransaction();
+
+                // Step A
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmInterprocessTest.java]: Same

File: realm/src/androidTest/java/io/realm/RealmInterprocessTest.java (hunk 4 above)
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmInterprocessTest.java]: Sorry, by bad. Just read the comment about the test notation. Keep it as it is.

File: realm/src/androidTest/java/io/realm/services/InterprocessService.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+import io.realm.Realm;
+import io.realm.entities.AllTypes;
+
+/**
+ * Helper service for multi-processes support testing.
+ */
+public class InterprocessService extends Service {
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/services/InterprocessService.java]: Perhaps rename this to RemoteProcessService? that indicates better that it is the "other" process and not the "main" one IMO

File: realm/src/androidTest/java/io/realm/RealmInterprocessTest.java [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+        }
+    }
+
+    // Helper handler to make it easy to interact with service process.
+    // This handler run in newly created thread's Looper. When writing the test case, remember to call
+    // Looper.loop() to start handling message.
+    // Pass the first thing you want to run to the constructor which will be posted to the beginning
+    // of the message queue.
+    // Write the comments of the test case like this:
+    // A-Z means steps running from service process.
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmInterprocessTest.java]: Perhaps move this instruction to the top of the file. At least I missed it on the first pass.

File: realm/src/androidTest/java/io/realm/RealmInterprocessTest.java [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+    // 1-9xx means steps running from the main process.
+    @SuppressLint("HandlerLeak") // SuppressLint bug, doesn't work
+    private class InterprocessHandler extends Handler {
+        // Timeout Watchdog. In case the service crashed or expected response is not returned.
+        // It is very important to feed the dog after the expected message arrived.
+        private final int timeout = 5000;
+        private boolean timeoutFlag = true;
```
Comment (nhachicha) [inline realm/src/androidTest/java/io/realm/RealmInterprocessTest.java]: `timeoutFlag` needs to be `volatile` it's accessed from different thread (_clearTimeoutFlag_ & _InterprocessHandler_)

File: realm/src/androidTest/java/io/realm/services/InterprocessService.java (hunk 7 above)
Comment (nhachicha) [inline realm/src/androidTest/java/io/realm/services/InterprocessService.java]: yep same for `serviceMessenger` could use the convention _main_  to indicate the main process & _remote_ to indicate the distant service

File: realm/src/androidTest/java/io/realm/RealmInterprocessTest.java [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+        private Runnable timeoutRunnable = new Runnable() {
+            @Override
+            public void run() {
+                if (timeoutFlag) {
+                    assertTrue("Timeout happened", false);
```
Comment (nhachicha) [inline realm/src/androidTest/java/io/realm/RealmInterprocessTest.java]: fancy way to crash :)

File: realm/src/androidTest/java/io/realm/RealmInterprocessTest.java (hunk 9 above)
Comment (nhachicha) [inline realm/src/androidTest/java/io/realm/RealmInterprocessTest.java]: could be renamed to `isTimeout`, `isTimeoutSet` for example

File: realm/src/androidTest/java/io/realm/RealmInterprocessTest.java [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+    private CountDownLatch serviceLatch;
+    private final ServiceConnection serviceConnection = new ServiceConnection() {
+        @Override
+        public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
+            serviceMessenger = new Messenger(iBinder);
+            serviceLatch.countDown();
+        }
+
+        @Override
+        public void onServiceDisconnected(ComponentName componentName) {
```
Comment (nhachicha) [inline realm/src/androidTest/java/io/realm/RealmInterprocessTest.java]: `serviceLatch = null;` for the next test?

File: realm/src/androidTest/java/io/realm/RealmInterprocessTest.java [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+                    latch.countDown();
+                }
+            }
+        });
+
+        thread.start();
+        try {
+            latch.await();
+        } catch (InterruptedException e) {
+            e.printStackTrace();
```
Comment (nhachicha) [inline realm/src/androidTest/java/io/realm/RealmInterprocessTest.java]: if you're Interrupted you should probably fail the test `throw e`

File: realm/src/androidTest/java/io/realm/RealmInterprocessTest.java (hunk 3 above)
Comment (nhachicha) [inline realm/src/androidTest/java/io/realm/RealmInterprocessTest.java]: why not also try a `android.os.Process.killProcess(remotePid);`

**Comment by nhachicha** [comment]
I think this looks solid, a bit complex to follow but the overall idea looks neat.
looking fwd to see how can we generalise this mini fwk outside #1328 use case (looking at your [Cursor](http://developer.android.com/reference/android/database/Cursor.html))
:+1:  some small comment s though

**Comment by beeender** [comment]
@realm/java Help to re-check. thanks!

File: realm/src/androidTest/java/io/realm/RealmInterprocessTest.java [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import io.realm.entities.AllTypes;
+import io.realm.services.RemoteProcessService;
+
+// This is built for testing multi processes related cases.
+// To build a test case, create an InterprocessHandler in your test case. This handler will run in newly created
```
Comment (emanuelez) [inline realm/src/androidTest/java/io/realm/RealmInterprocessTest.java]: This handler will run in **the** newly created

File: realm/src/androidTest/java/io/realm/RealmInterprocessTest.java [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+// thread's Looper. Remember to call Looper.loop() to start handling message.
```
Comment (emanuelez) [inline realm/src/androidTest/java/io/realm/RealmInterprocessTest.java]: message**s**

File: realm/src/androidTest/java/io/realm/RealmInterprocessTest.java [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+            serviceStartLatch = null;
+        }
+    };
+
+    // It is necessary to overload this method.
+    // AndroidTestRunner does call Looper.prepare() and we can have a looper in the case. The problem is all the test
+    // cases are running in a single thread!!! And after Looper.quit() called, it cannot start again. That means we
+    // can only have one case in this class LoL.
+    // By overloading this method, we create a new thread and looper to run the real case. And use latch to wait until
+    // it is finished. Then we can get rid of creating thread in the test method, using array to store exception, many
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmInterprocessTest.java]: creating *the* thread

File: realm/src/androidTest/java/io/realm/RealmInterprocessTest.java [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+            final int servicePid = getServiceInfo().pid;
+
+            @Override
+            public void handleMessage(Message msg) {
+                super.handleMessage(msg);
+                if (msg.what == RemoteProcessService.stepExitProcess_A.message) {
+                    // Step 1
+                    clearTimeoutFlag();
+                    try {
+                        // Timeout is 5 seconds. 3 (6x500ms) second should be enough to quit the process.
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmInterprocessTest.java]: second -> seconds

**Comment by cmelchior** [comment]
Apart from some spelling this looks good :+1:

File: realm/src/androidTest/java/io/realm/services/RemoteProcessService.java [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+
+    @Override
+    public boolean onUnbind(Intent intent) {
+        stopSelf();
+        return super.onUnbind(intent);
+    }
+
+    @Override
+    public int onStartCommand(Intent intent, int flags, int startId) {
+        return  START_NOT_STICKY;
```
Comment (emanuelez) [inline realm/src/androidTest/java/io/realm/services/RemoteProcessService.java]: one space only

Comment (beeender) [inline realm/src/androidTest/java/io/realm/services/RemoteProcessService.java]: How can you see that??!!

**Comment by beeender** [comment]
retest this please

**Comment by cmelchior** [comment]
:+1:

**Comment by emanuelez** [comment]
:+1:

**Comment by beeender** [comment]
retest this please

**Comment by emanuelez** [comment]
retest this please

**Pull request merged**